src/scr_flush_file.c

```diff
@@ -40,6 +45,167 @@
+  kvtree* summary;
+
+  flush_key_dataset = kvtree_get_kv_int(flush_file, SCR_FLUSH_KEY_DATASET, dataset_id);
+  if (flush_key_dataset == NULL) {
+    printf("%s: no flush file entry for dataset %d\n", __func__, dataset_id);
+    return 1; /* No entry for this dataset */
+  }
+
+  summary = kvtree_get(flush_key_dataset, SCR_FLUSH_KEY_DSETDESC);
+
+  asprintf(&summary_file, "%s/.scr/scr.dataset.%d/summary.scr", prefix, dataset_id);
+
+  rc = scr_flush_summary_file(summary, 1, summary_file);
+
+  scr_free(&summary_file);
+
+  /* Update the flush file that we're not longer transferring */
+  scr_flush_file_location_unset_with_path(dataset_id,
+    SCR_FLUSH_KEY_LOCATION_SYNC_FLUSHING, flush_file_path);
+
+  return rc;
+}
+
+/*
+ * Given a path to a state_file, resume and finalize all transfers for all
+ * files in the state_file.
+ */
+int resume_transfer(char* state_file_path)
+{
+  int id;
+  int rc;
+  AXL_Init();
+
+  id = AXL_Create(AXL_XFER_STATE_FILE, "scr", state_file_path);
+  if (id < 0) {
+    scr_err("Error: AXL_Create() = %d\n", id);
+    rc = id;
+    goto end;
+  }
+
+  rc = AXL_Resume(id);
+  if (rc != AXL_SUCCESS) {
+    scr_err("Error: AXL_Resume(%d) = %d\n", id, rc);
+    goto end;
+  }
+
+  rc = AXL_Wait(id);
+  if (rc != AXL_SUCCESS) {
+    scr_err("Error: AXL_Wait(%d), = %d\n", id, rc);
+    goto end;
+  }
+
+  rc = AXL_Free(id);
+  if (rc != AXL_SUCCESS) {
+    scr_err("Error: AXL_Free(%d), = %d\n", id, rc);
+    goto end;
+  }
+
+  rc = AXL_Finalize();
+  if (rc != AXL_SUCCESS) {
+    scr_err("Error: AXL_Finalize(), = %d\n", id, rc);
+  }
+end:
+  return rc;
+}
+
+/*
+ * Resume and wait for any previous transfers to complete, and finalize them.
+ *
+ * This only resumes/waits for the AXL transfers to complete.  It does not
+ * update SCR's flush file nor write the summary file.
+ *
+ * Returns 0 on success, non-zero otherwise.
+ */
+int resume_transfers(char* prefix, int dataset_id)
+{
+  char* rank2file_path = NULL;
+  int rc;
+  int ranks;
+  char num_str[21]; /* Max number in uint64_t is 20 digits + '\0' */
+  unsigned long i;
+  char* file;
+  char* state_file;
+  kvtree* ranks_tree;
+  kvtree* rank_subtree;
+
+  ranks_tree = kvtree_new();
+
+  asprintf(&rank2file_path, "%s/.scr/scr.dataset.%d/rank2file", prefix, dataset_id);
+
+  rc = kvtree_read_scatter_single(rank2file_path, ranks_tree);
+  if (rc) {
+    scr_err("kvtree_read_scatter_single(%s) = %d\n", __func__, rank2file_path, rc);
+    return rc;
+  }
+
+  /*
+   * 'ranks_tree' is a kvtree that looks like:
+   *
+   *    79
+   *      FILE
+   *        ckpt.1/rank_79.ckpt
+   *    73
+   *      FILE
+   *        ckpt.1/rank_73.ckpt
+   *    74
+   *      FILE
+   *        ckpt.1/rank_74.ckpt
+   */
+  ranks = kvtree_size(ranks_tree);
+  for (i = 0; i < ranks; i++) {
+   /*
+    * Get each:
+    *
+    *    74
+    *      FILE
+    *        ckpt.1/rank_74.ckpt
+    *...
+    */
+    sprintf(num_str, "%lu", i);
+    rank_subtree =  kvtree_get(ranks_tree, num_str);
```

Comment:
Let's fuse this if getf works and then drop num_str:
```suggestion
    rank_subtree =  kvtree_getf(ranks_tree, "%lu", i);
```